accepted/2020/linking-libraries.md

```diff
@@ -0,0 +1,152 @@
+
+ILLink is capable of saving the dependency graph when it trims an application. There is a tool, [Linker Analyzer](https://github.com/mono/linker/tree/d3eb4ea8179516b0554a3d3089c71062496bd6f0/src/analyzer), that can load this graph and allows a developer to analyze it. Example questions that can be answered are:
+
+1. What are the biggest types/methods left after trimming?
+2. What is the call chain that is keeping type/method `foo` from being trimmed?
+
+Today the Linker Analyzer is a console application. One enhancement we can make is to put the core logic into a NuGet package, and allow a developer to analyze their application using the [.NET Interactive](https://github.com/dotnet/interactive/blob/master/README.md) notebook experience. Advantages of this experience are:
+
+1. The dependency graph can be large (many MBs). Loading and unloading it to ask one question at a time in a console app takes time. In a notebook, it is loaded into memory once and cached. Many questions can be asked faster.
+2. Developers can write their own logic and extensions to the dependency graph.
+3. We can have template notebooks that show examples of how to analyze an application's size.
+4. When a new feature is added to the Analyzer, we just need to create a public API. We don't need to create command line arguments.
+5. We have the option of visuals and more interactive output - for example hyperlinks. In a console application, you are stuck with text.
+
+## Testing
+
+Testing the linked applications falls into two buckets:
+1. Does the application still work?
+2. Is the application smaller?
+
+In general, using unit tests to ensure a linked application still works isn't a great testing strategy. Unit tests, by definition, try exercising all the code in the library. If all the code in the library is being exercised, the linker won't be able to trim any of it. Thus if nothing is trimmed, you aren't confirming that the library still works when not all of the code is present.
+
+Instead of using unit tests, we should create dedicated applications to test that the libraries are linker safe. One set of applications will be higher-level applications that we want to measure the size over time. Another set will target specific APIs that needed to be annotated to ensure the annotations and the linker are working correctly.
+
+As we are measuring size over time, we would not like to get "noise" in the sizes to cause false positives. For example, if an application size gets larger by ~10 KB, we don't necessarily want the test to fail. And conversely, if an application size suddenly get 50% larger (or smaller), we probably do want to alert someone - as an anomaly was detected and it is possibly a bug somewhere.
```

Comment:
@billwert is working on this and has design/proposal for this kind of testing. It might be better to link it from here instead.

Reply:
@billwert - I haven't seen your proposal, can you share a link that I can add?